### PR TITLE
Ensure callable functions have invoker correctly set to "public"

### PR DESCRIPTION
**Pull request by chrisbrown-io**
### Description

Implemented fix for issue introduced in `v10.3.0` where callable functions were not having their `invoker` property correctly set to `"public"` that would in turn cause access issues with the deployed functions (as detailed in #4327).

Will also fix issues such as #3965, that are presumably cause by version mismatches/upgrades introducing this issue.

[See comment for explicit info](https://github.com/firebase/firebase-tools/issues/4327#issuecomment-1072650006).

Fixes issue by adding specific handling for callable functions and always setting the `invoker` to `"public"`.

Also fixed issue with `taskQueueTrigger` endpoints calling wrong method/passing wrong args when trying to set invoker.

### Scenarios Tested

Thanks to @taeold we have explicit tests for each function type and the expected outcomes for both `createV1Function` and `createV2Function`  methods.

### CODE REVIEW

**Comment by chrisbrown-io** [comment]
@taeold - you mentioned in your [previous comment](https://github.com/firebase/firebase-tools/issues/4327#issuecomment-1072690202) that the invoker is not updated on update, however the source code/tests for the update methods look as if they would update the invoker if set (and subsequently not update the invoker for callable functions unless additional logic was added).

Are you saying we don't need to handle this case, as users can't specify an `invoker` for callable functions, meaning this scenario of changing the invoker on update will never happen?

File: src/deploy/functions/release/fabricator.ts [hunk 1]
```diff
@@ -233,8 +233,16 @@ export class Fabricator {
           })
           .catch(rethrowAs(endpoint, "set invoker"));
       }
+    } else if (backend.isCallableTriggered(endpoint)) {
+      // Callable functions should always be public
+      const invoker = ["public"];
```
Comment (taeold) [inline src/deploy/functions/release/fabricator.ts:238]: nit* Personal preference, but I'd inline this value instead of creating an intermediate var.

Same comment for L310

File: src/deploy/functions/release/fabricator.ts [hunk 2]
```diff
@@ -233,8 +233,16 @@ export class Fabricator {
...
+      await this.executor
+        .run(async () => {
+          await gcf.setInvokerCreate(endpoint.project, backend.functionName(endpoint), invoker);
+        })
+        .catch(rethrowAs(endpoint, "set invoker"));
     } else if (backend.isTaskQueueTriggered(endpoint)) {
-      // Like HTTPS triggers, taskQueueTriggers have an invoker, but unlike HTTPS they don't default
+      // Like other triggers, taskQueueTriggers can have an invoker, but they don't default
```
Comment (taeold) [inline src/deploy/functions/release/fabricator.ts:245]: The previous comment is still true since only task queue triggers and http triggers can explicitly set invokers. I'd prefer if we revert this change.

Same comment for L315

**Review by taeold — CHANGES_REQUESTED** [review]
> Are you saying we don't need to handle this case, as users can't specify an invoker for callable functions, meaning this scenario of changing the invoker on update will never happen?

Yes. User may have changed invoker for good reasons. I think we should respect that.

Also - can we add an entry to CHANGELOG.md file?

File: src/deploy/functions/release/fabricator.ts (hunk 1 above)
Comment (chrisbrown-io) [reply src/deploy/functions/release/fabricator.ts:238]: Sure - makes sense.

File: src/deploy/functions/release/fabricator.ts (hunk 2 above)
Comment (chrisbrown-io) [reply src/deploy/functions/release/fabricator.ts:245]: 👍

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1 @@
+- Fix for issue introduced where callable functions were not having their invoker property correctly set to "public" (#4327).
```
Comment (taeold) [inline CHANGELOG.md:1]:
```suggestion
- Fixes bug where new callable functions were not publicly accessible (#4327).
```

**Review by taeold — APPROVED** [review]
🔥 Amazing contribution!
